### PR TITLE
[Merged by Bors] - feat: `Pairwise` and composition with an equiv

### PR DESCRIPTION
**Pull request by jsm28**
Add a lemma about `Pairwise` being preserved by composing the arguments of the pairwise relation with an equiv.

```lean
lemma EquivLike.pairwise_comp_iff {X : Type*} {Y : Type*} {F} [EquivLike F Y X]
    (f : F) (p : X → X → Prop) : Pairwise (fun y z ↦ p (f y) (f z)) ↔ Pairwise p := by
```

This depends on `Mathlib.Data.FunLike.Equiv` and
`Mathlib.Logic.Pairwise`, neither of which imports the other, so to avoid increasing the imports of very basic files I put it in its own new file, but feel free to suggest a better location of this lemma (and to golf the proof).

From AperiodicMonotilesLean.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by jsm28** [comment]
To consider in review: do we also want variants for unbundled bijective functions, and one-way implications for unbundled injective and surjective functions and for `EmbeddingLike`, and where should all those variants go if so?

File: Mathlib/Logic/Equiv/Pairwise.lean [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+-/
+import Mathlib.Data.FunLike.Equiv
+import Mathlib.Logic.Pairwise
+
+/-!
+# Interaction of equivalences with `Pairwise`
+-/
+
+lemma EquivLike.pairwise_comp_iff {X : Type*} {Y : Type*} {F} [EquivLike F Y X]
+    (f : F) (p : X → X → Prop) : Pairwise (fun y z ↦ p (f y) (f z)) ↔ Pairwise p := by
```
Comment (eric-wieser) [inline Mathlib/Logic/Equiv/Pairwise.lean:14]: Maybe
```suggestion
    (f : F) (p : X → X → Prop) : Pairwise (p on f) ↔ Pairwise p := by
```

File: Mathlib/Logic/Equiv/Pairwise.lean [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  refine ⟨fun h ↦ fun i j hij ↦ ?_,
```
Comment (eric-wieser) [inline Mathlib/Logic/Equiv/Pairwise.lean:15]: I think there might be a shorter proof via `Surjective.forall`

**Comment by jsm28** [comment]
I've now added the lemma for unbundled functions (in `Mathlib.Logic.Pairwise`, since they don't have any further dependencies), and one for `EmbeddingLike` (in the same new file as the `EquivLike` one; the question about the best locations still applies there, for both lemmas, and quite likely the two belong in different files). I've also used `Function.Surjective.forall₂` for golfing as suggested.

File: Mathlib/Logic/Equiv/Pairwise.lean (hunk 2 above)
Comment (jsm28) [reply Mathlib/Logic/Equiv/Pairwise.lean]: Now using `Function.Surjective.forall₂` (in the proof of one of the new unbundled lemmas).

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**